runners/java-fn-execution/src/main/java/org/apache/beam/runners/fnexecution/environment/DockerEnvironmentFactory.java

```diff
@@ -152,8 +152,10 @@ public RemoteEnvironment createEnvironment(Environment environment) throws Excep
       LOG.debug("Created Docker Container with Container ID {}", containerId);
       // Wait on a client from the gRPC server.
       while (instructionHandler == null) {
+        Preconditions.checkArgument(
+            docker.isContainerRunning(containerId), "No container running for id " + containerId);
```

Comment:
Can we assume that the container startup is synchronous? If not, would it make sense to add a retry logic here?

Reply:
The main issue that I observed here was when the container failed to start. As it is started in detached mode, we don't know the state of the container and simply wait on the control client to connect.

Synchronous container start will not work and will be very difficult to orchestrate.

Reply:
The check is too strict then. We won't get `RUNNING` directly after starting the container. We need something like a repeated check which times out after some max startup time.

Reply:
I am not sure if the 'docker run -d <image> ' return before the container is in running state. From running state, it can dies because of user code error/completion.

However it makes sense to check the state after the initial suggested wait of 1 min.